collectors/metrics/cmd/metrics-collector/main.go

```diff
@@ -554,68 +537,241 @@ func initConfig(o *Options) (error, *forwarder.Config) {
+					URL:       from,
+					QueryURL:  fromQuery,
+					Token:     o.FromToken,
+					TokenFile: o.FromTokenFile,
+					CAFile:    o.FromCAFile,
+				},
+				StatusClient: statusClient,
+				ToClientConfig: forwarder.ToClientConfig{
+					URL:      toUpload,
+					CAFile:   o.ToUploadCA,
+					CertFile: o.ToUploadCert,
+					KeyFile:  o.ToUploadKey,
+				},
+				AnonymizeLabels:         o.AnonymizeLabels,
+				AnonymizeSalt:           o.AnonymizeSalt,
+				AnonymizeSaltFile:       o.AnonymizeSaltFile,
+				Debug:                   o.Verbose,
+				Interval:                o.Interval,
+				EvaluateInterval:        o.EvaluateInterval,
+				LimitBytes:              o.LimitBytes,
+				Matchers:                shard,
+				Transformer:             transformer,
+				Logger:                  log.With(o.Logger, "shard", i),
+				SimulatedTimeseriesFile: o.SimulatedTimeseriesFile,
+			}
+		}
+		return shardCfgs, nil
+	default:
+		return nil, errors.New("invalid agent type")
 	}
 }
 
-// serveLastMetrics retrieves the last set of metrics served.
-func serveLastMetrics(l log.Logger, worker *forwarder.Worker) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		if req.Method != "GET" {
-			w.WriteHeader(http.StatusMethodNotAllowed)
-			return
+func runMultiWorkers(o *Options, cfg *forwarder.Config) error {
+	for i := 1; i < int(o.WorkerNum); i++ {
+		opt := &Options{
+			From:                    o.From,
+			FromQuery:               o.FromQuery,
+			ToUpload:                o.ToUpload,
+			FromCAFile:              o.FromCAFile,
+			FromTokenFile:           o.FromTokenFile,
+			ToUploadCA:              o.ToUploadCA,
+			ToUploadCert:            o.ToUploadCert,
+			ToUploadKey:             o.ToUploadKey,
+			Matchers:                o.Matchers,
+			RecordingRules:          o.RecordingRules,
+			Interval:                o.Interval,
+			Labels:                  map[string]string{},
+			SimulatedTimeseriesFile: o.SimulatedTimeseriesFile,
+			Logger:                  o.Logger,
+			DisableHyperShift:       o.DisableHyperShift,
+			DisableStatusReporting:  o.DisableStatusReporting,
 		}
-		families := worker.LastMetrics()
-		protoTextFormat := expfmt.NewFormat(expfmt.TypeProtoText)
-		w.Header().Set("Content-Type", string(protoTextFormat))
-		encoder := expfmt.NewEncoder(w, protoTextFormat)
-		for _, family := range families {
-			if family == nil {
-				continue
+		for _, flag := range o.LabelFlag {
+			values := strings.SplitN(flag, "=", 2)
+			if len(values) != 2 {
+				return fmt.Errorf("--label must be of the form key=value: %s", flag)
 			}
-			if err := encoder.Encode(family); err != nil {
-				logger.Log(l, logger.Error, "msg", "unable to write metrics for family", "err", err)
-				break
+			if values[0] == "cluster" {
+				values[1] += "-" + fmt.Sprint(i)
+			}
+			if values[0] == "clusterID" {
+				values[1] = string(uuid.NewUUID())
 			}
+			opt.Labels[values[0]] = values[1]
```

Comment:
`opt.Labels` already exists at this stage no? It's been generated in `initShardedConfigs`. We could iterate directly on it. And why are we setting the `cluster` and `clusterID` at this stage and not in the other function?
I might be lost

Reply:
So this runMultiWorkers function isn't really used anywhere in main flow, this is part of the simulation testing flow. I just moved the function, so it diffs this way, but I haven't actually modified anything in this function,

https://github.com/stolostron/multicluster-observability-operator/blob/main/collectors/metrics/cmd/metrics-collector/main.go#L442

Reply:
As for why it is the way it is, I don't really know, but I wanted to remove this simulation testing, it is mixed in with business logic and is not really a great simulation of anything. Can do in later PRs

Reply:
Ack no worries, I did not pay too much attention between what was already there, and what you added. We can leave it as is.